src/Compilers/CSharp/Portable/BoundTree/BoundNodeExtensions.cs

```diff
@@ -94,5 +97,163 @@ private class ContainsAwaitVisitor : BoundTreeWalkerWithStackGuardWithoutRecursi
+                }
+
+                switch (current.Right)
+                {
+                    case BoundBinaryOperator rightOperator:
+                        binaryOperatorCallback?.Invoke(rightOperator, arg);
+                        pushLeftNodes(rightOperator, stack);
+                        break;
+                    case TInterpolatedStringType interpolatedString:
+                        if (!visitor(interpolatedString, arg))
+                        {
+                            return false;
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(current.Right.Kind);
+                }
+            }
+
+            Debug.Assert(stack.Count == 0);
+            stack.Free();
+            return true;
+
+            static void pushLeftNodes(BoundBinaryOperator binary, ArrayBuilder<BoundBinaryOperator> stack)
+            {
+                Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString) || binary.IsUnconvertedInterpolatedStringAddition);
+                BoundBinaryOperator? current = binary;
+                while (current != null)
+                {
+                    stack.Push(current);
+                    current = current.Left as BoundBinaryOperator;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Rewrites a BoundBinaryOperator composed of interpolated strings (either converted or unconverted) iteratively, without
+        /// recursion.
+        /// </summary>
+        /// <param name="binary">The original top of the binary operations.</param>
+        /// <param name="arg">The callback args.</param>
+        /// <param name="interpolatedStringFactory">
+        /// Rewriter for the BoundInterpolatedString or BoundUnconvertedInterpolatedString parts of the binary operator. Passed the callback
+        /// parameter, the original interpolated string, and the index of the interpolated string in the tree.
+        /// </param>
+        /// <param name="binaryOperatorFactory">
+        /// Rewriter for the BoundBinaryOperator parts fo the binary operator. Passed the callback parameter, the original binary operator, and
+        /// the rewritten left and right components.
+        /// </param>
+        public static TResult RewriteInterpolatedStringAddition<TArg, TInterpolatedStringType, TResult>(
+            this BoundBinaryOperator binary,
+            TArg arg,
+            Func<TArg, TInterpolatedStringType, int, TResult> interpolatedStringFactory,
+            Func<TArg, BoundBinaryOperator, TResult, TResult, TResult> binaryOperatorFactory)
+            where TInterpolatedStringType : BoundExpression
+        {
+            Debug.Assert(typeof(TInterpolatedStringType) == typeof(BoundUnconvertedInterpolatedString) || typeof(TInterpolatedStringType) == typeof(BoundInterpolatedString));
```

Comment:
Might it be possible to remove the `TInterpolatedStringType` type parameter and use `BoundInterpolatedStringBase` instead? It's abstract and only has those two derived types. 
(same question applies to `VisitBinaryOperatorInterpolatedString`)

Reply:
I'll tighten the constraint, but I'd prefer to keep the type parameter to ensure that we're being specific about the allowed things and fail when we encounter something we didn't expect.